### ISSUE TITLE
Strange behavior of specs

### ISSUE THREAD

**Issue by Codcore** [ysbaddaden/minitest.cr#4]
Decided finally to use your library for testing. When running spec file several times, each time I got different results:

```
rystal-projects/Amethyst] - [Tue Jun 16, 04:53]
└─[$] <git:(mini-test*)> cr spec/base/application_spec.cr


Finished in 3.69 milliseconds
0 examples, 0 failures, 0 errors, 0 pending
........

Finished in 00:00:00.0013290, 752.445 runs/s

8 tests, 0 failures, 0 errors, 0 skips
┌─[kavachaj@Rails] - [~/crystal-projects/Amethyst] - [Tue Jun 16, 04:55]
└─[$] <git:(mini-test*)> cr spec/base/application_spec.cr


Finished in 3.71 milliseconds
0 examples, 0 failures, 0 errors, 0 pending
.FF.....

Finished in 00:00:00.0064350, 155.4 runs/s

  1) Failure:
Base::App::initialize method#test_should_set_app_directory [spec/base/application_spec.cr:15]:
Expected "" but got "/home/kavachaj/crystal-projects/Amethyst/spec/base"

  2) Failure:
Base::App::initialize method#test_should_set_default_middleware [spec/base/application_spec.cr:19]:
Expected #<Amethyst::Middleware::MiddlewareStack:0x8fe9f40 @middlewares=[TestMiddleware]> to include Amethyst::Middleware::ShowExceptions

8 tests, 2 failures, 0 errors, 0 skips
Program terminated abnormally with error code: 65280
┌─[kavachaj@Rails] - [~/crystal-projects/Amethyst] - [Tue Jun 16, 04:55]
└─[$] <git:(mini-test*)> cr spec/base/application_spec.cr


Finished in 3.72 milliseconds
0 examples, 0 failures, 0 errors, 0 pending
F.......

Finished in 00:00:00.0046380, 215.61 runs/s

  1) Failure:
Base::App::initialize method#test_should_set_default_middleware [spec/base/application_spec.cr:19]:
Expected #<Amethyst::Middleware::MiddlewareStack:0x9897f40 @middlewares=[TestMiddleware]> to include Amethyst::Middleware::ShowExceptions

8 tests, 1 failures, 0 errors, 0 skips
Program terminated abnormally with error code: 65280
┌─[kavachaj@Rails] - [~/crystal-projects/Amethyst] - [Tue Jun 16, 04:55]
└─[$] <git:(mini-test*)> cr spec/base/application_spec.cr


Finished in 3.69 milliseconds
0 examples, 0 failures, 0 errors, 0 pending
........

Finished in 00:00:00.0013280, 753.012 runs/s

8 tests, 0 failures, 0 errors, 0 skips
```

But code has not changed! Maybe it is because of concurrency?I created separate branch of repository - https://github.com/Codcore/Amethyst/tree/mini-test
You can test it yourself (testing code is here - https://github.com/Codcore/Amethyst/blob/mini-test/spec/base/application_spec.cr)

**Comment by ysbaddaden** [comment]
It may have something to do with concurrency, yes. Tests no longer run in parallel anymore since they now run in coroutines, and no longer in threads, yet, code _and_ tests must be resilient to concurrency and be prepared for coroutines to have already run or not before assertions are executed.

You may want to try with `--parallel=1` and see if the problem still happens. If not, then, minitest helped you discover a concurrency issue :smile:

**Comment by ysbaddaden** [comment]
I flagged as "question" until we identify a bug in minitest.

**Comment by Codcore** [comment]
@ysbaddaden but when I change specs to run with Crystal standart spec module, they pass well everytime. Tried --parallel=1:

```
cr spec/base/application_spec.cr --parallel=1   
Error: Invalid option: --parallel=1
```

**Comment by ysbaddaden** [comment]
Ah sorry, you must use `--` to separate arguments sent to the generated binary, otherwise they are interpreted by crystal:

```
$ crystal spec/base/application_spec.cr -- --parallel=1
```

You can list available args:

```
$ crystal spec/base/application_spec.cr -- --help
```

**Comment by Codcore** [comment]
@ysbaddaden 

```
crystal spec/base/application_spec.cr -- --parallel=1
```

Output:

```
Invalid option: --parallel=1 (OptionParser::InvalidOption)
*OptionParser::Exception@Exception#initialize<OptionParser::InvalidOption, String>:Array(String) +71 [0]
*OptionParser::InvalidOption#initialize<OptionParser::InvalidOption, String>:Array(String) +119 [0]
*OptionParser::InvalidOption::new<String>:OptionParser::InvalidOption +133 [0]
*OptionParser::ParseTask#check_invalid_options<OptionParser::ParseTask>:Array(String)? +360 [0]
*OptionParser::ParseTask#parse<OptionParser::ParseTask>:Array(String)? +252 [0]
*OptionParser#parse<OptionParser, Array(String)>:Array(String)? +160 [0]
__crystal_main +24712 [0]
main +148 [0]
__libc_start_main +247 [0]
_start +33 [0]
Program terminated abnormally with error code: 256
```

**Comment by ysbaddaden** [comment]
Damn. It looks like OptionParser doesn't support the `--option=value` syntax. This will work (I just tested on computer, instead of lazily replying from my phone):

```
$ crystal spec/base/application_spec.cr -- --parallel 1
```

**Comment by Codcore** [comment]
@ysbaddaden strange, but it doesn't work on my PC

```
crystal spec/base/application_spec.cr -- --parallel 1
Invalid option: --parallel (OptionParser::InvalidOption)
*OptionParser::Exception@Exception#initialize<OptionParser::InvalidOption, String>:Array(String) +71 [0]
*OptionParser::InvalidOption#initialize<OptionParser::InvalidOption, String>:Array(String) +119 [0]
*OptionParser::InvalidOption::new<String>:OptionParser::InvalidOption +133 [0]
*OptionParser::ParseTask#check_invalid_options<OptionParser::ParseTask>:Array(String)? +360 [0]
*OptionParser::ParseTask#parse<OptionParser::ParseTask>:Array(String)? +252 [0]
*OptionParser#parse<OptionParser, Array(String)>:Array(String)? +160 [0]
__crystal_main +24463 [0]
main +148 [0]
__libc_start_main +247 [0]
_start +33 [0]
Program terminated abnormally with error code: 256
```

Available options are:

```
 crystal spec/base/application_spec.cr --help         
Usage: crystal run [options] [programfile] [--] [arguments]

Options:
    -d, --debug                      Add symbolic debug info
    -D FLAG, --define FLAG           Define a compile-time flag
    --emit [asm|llvm-bc|llvm-ir|obj] Comma separated list of types of output for the compiler to emit
    -h, --help                       Show this message
    --ll                             Dump ll to .crystal directory
    --link-flags FLAGS               Additional flags to pass to the linker
    --mcpu CPU                       Target specific cpu type
    --no-color                       Disable colored output
    --no-build                       Disable build output
    -o                               Output filename
    --prelude                        Use given file as prelude
    --release                        Compile in release mode
    -s, --stats                      Enable statistics output
    --single-module                  Generate a single LLVM module
    --threads                        Maximum number of threads to use
    --verbose                        Display executed commands
```

I tried `crystal spec/base/application_spec.cr --threads 1`, but specs behave in the same way :(

**Comment by ysbaddaden** [comment]
You're missing a `--` separator again in your calls to `--help`, but the first call with `-- --parallel 1` should be fine. Actually I tried and `-- --parallel=1` is fine too :sob: 

I'm running the released Crystal v0.7.3.
